From a755a6f7d89a0db7f520b2fcef33b623bb2e4b25 Mon Sep 17 00:00:00 2001
From: Kurt Newman <kurt.newman@cpanel.net>
Date: Wed, 23 Apr 2014 16:59:28 -0500
Subject: [PATCH 1/2] Concurrent logging patch to fix setuid Apache

Case 76493: Resolve locking mutex conflicts with Apache modules
that log as different users.  Serial logging can't be used to
create or lock mutexes because subsequent calls as different
users cause file ownership problems.

By using concurrent logging (one file per logged transaction),
and adding a per-username directory path (via the patch), we
get around the locking issue because each user writes to its
own directory.

This is similar to how EAccelerator caches files on disk.
---
 apache2/msc_logging.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/apache2/msc_logging.c b/apache2/msc_logging.c
index 3323fac..c916ba0 100644
--- a/apache2/msc_logging.c
+++ b/apache2/msc_logging.c
@@ -224,11 +224,18 @@ static char *construct_auditlog_filename(apr_pool_t *mp, const char *uniqueid) {
     apr_time_exp_t t;
     char tstr[300];
     apr_size_t len;
+    apr_uid_t uid;
+    apr_gid_t gid;
+    char *username;
 
     apr_time_exp_lt(&t, apr_time_now());
 
     apr_strftime(tstr, &len, 299, "/%Y%m%d/%Y%m%d-%H%M/%Y%m%d-%H%M%S", &t);
-    return apr_psprintf(mp, "%s-%s", tstr, uniqueid);
+
+    apr_uid_current(&uid, &gid, mp);
+    apr_uid_name_get(&username, uid, mp);
+
+    return apr_psprintf(mp, "/%s%s-%s", username, tstr, uniqueid);
 }
 
 /**
-- 
2.2.0

